consensus/istanbul/backend/backend.go

```diff
@@ -1005,6 +1008,18 @@ func (sb *Backend) RemoveProxy(node *enode.Node) error {
 	}
 }
 
+func (sb *Backend) RetrieveUptimeScoreBuilder(header *types.Header, epochSize uint64, lookbackWindowFn func() uint64) uptime.Builder {
+	epoch := istanbul.GetEpochNumber(header.Number.Uint64(), epochSize)
+
+	if sb.uptimeMonitor == nil || sb.uptimeMonitor.GetEpoch() != epoch {
+		valSet := sb.GetValidators(header.Number, header.Hash())
+		builder := uptime.NewMonitor(epochSize, epoch, lookbackWindowFn(), len(valSet))
```

Comment:
Because we call a newMonitor every time, if we have it, we create it, otherwise we return the instance that we have.
As I don't actually know if the loockbackWindow, was cached, I basically sent it as a function. If the monitor instance exist, I will return it, if it don't, I will create the monitor, and invoke this function. So in a way, I'm calling it once per epoch, without relying if it is cached or not

Reply:
Oh, and little detail. For the validation set, it does not require the state, but for the lookback window, we need it.
We can change that to cache it, but I saw this as a simpler solution